### ISSUE TITLE
Trim adapter sequence from reads

### ISSUE THREAD

**Issue by donkirkby** [cfe-lab/MiCall#312]
Add the [cutadapt](https://cutadapt.readthedocs.io) tool to the pipeline.

> Cutadapt finds and removes adapter sequences, primers, poly-A tails and other types of unwanted sequence from your high-throughput sequencing reads.
> 
> Cleaning your data in this way is often required: Reads from small-RNA sequencing contain the 3’ sequencing adapter because the read is longer than the molecule that is sequenced. Amplicon reads start with a primer sequence. Poly-A tails are useful for pulling out RNA from your sample, but often you don’t want them to be in your reads.

* [x] Add cutadapt and try it on some samples.
* [x] See if multiple versions of cutadapt can be installed side by side with an externals wrapper.

**Comment by veratai** [comment]
example use of cutadapt for libraries prepared with TruSeq adapters: 

cutadapt -q 20,20 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT -m 50 -o fileR1.cutadapt.fastq -p fileR2.cutadapt.fastq fileR1.fastq fileR2.fastq

These options will work for TruSeq libraries, dual-indexed (and their equivalent):
-a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT

or this (which just uses the common parts to both adapters, may reduce headaches):

-a AGATCGGAAGAGC -A AGATCGGAAGAGC


For Nextera, dual-indexed, use this:

-a CTGTCTCTTATACACATCTCCGAGCCCACGAGAC   -A CTGTCTCTTATACACATCTGACGCTGCCGACGA 

or this should also work (using common parts of both adapters):

-a CTGTCTCTTATACACATCT -A CTGTCTCTTATACACATCT


In the lab, I’m pretty sure we mostly deal with TruSeq and Nextera adapters, but for the research folk, we are testing new kits which may have different adapters (unlikely, but should double-check).  Here’s the file with all Illumina “standard” adapters. See page 12 and 19. 

http://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences_1000000002694-01.pdf

**Comment by donkirkby** [comment]
Winnie gave me these adapter sequences to trim, retrieved from [Illumina].

Nextera XT :
Read 1
5’ TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG
Read 2
5’ GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG


NEB random primer or TruSeq:

5’ ACACTCTTTCCCTACACGACGCTCTTCCGATCT

5’ GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT

[Illumina]: http://support.illumina.com/downloads/illumina-customer-sequence-letter.html

**Comment by donkirkby** [comment]
I looked in sample 67516A-V3-1-V3LOOP_S15 from the 10 Nov 2016 run, and found the Nextera XT adapter sequences in many reads that were soft clipped at position 910 in HIV-env. However, the read 2 adapter appeared in the read 1 FASTQ as its reverse complement, and the read 1 adapter appeared in the read 2 FASTQ as its reverse complement.

I looked in sample 95973A-HCV_S3 from the 14 Nov 2016 run, and found the NEB random primer adapter in lots of unmapped reads, as well as some of the mapped reads. One example mapped read is M01841:268:000000000-ARJVW:1:1104:21554:4356. As with the other sample, the read 2 adapter appeared in the read 1 FASTQ as its reverse complement, and the read 1 adapter appeared in the read 2 FASTQ as its reverse complement.

**Comment by donkirkby** [comment]
I did my first test of trimming the adapters with sample 67516A-V3-1-V3LOOP_S15, but it hardly helped at all. It only allowed 27 more reads to map in the `prelim_map` step, and all of them just trimmed the last three bases. 24 were read 1, 3 were read 2.

1651 read 1 reads and 1616 read 2 reads contained a full adapter sequence, but removing the adapter sequences didn't let them map to the reference. I tried blasting some of the remaining sequence, and it mapped to a different HIV-env reference.

**Comment by donkirkby** [comment]
Trimming the adapters with sample 95973A-HCV_S3 was much more helpful. With only censoring, the `prelim_map` step mapped 576 reads, but with adapter trimming, it mapped 822 reads (a 43% increase). When I looked at 20 of the reads that were only able to map with trimming, they looked as expected. Shorter than 250 nucleotides, with both reads mapped to the same position. The shortest one I saw was 9 nucleotides.

Since that worked so well, I checked sample 73433A-HCV_S42 from the 10 Nov 2016 run to see if a Nextera sample would work better than the V3LOOP amplicon sample I tried before. It worked much better, going from 256914 reads to 494305 reads (a 92% increase). Again, I looked at 20 reads, and they all looked as expected.

Since this works so well on the Nextera samples, I'm going to commit the adapter trimming feature, and then look for other problems with the V3LOOP amplicon samples.
